power_grid_model_c/power_grid_model/include/power_grid_model/component/current_sensor.hpp

```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+  protected:
+    double convert_direction() const {
+        if (terminal_type_ == MeasuredTerminalType::load || terminal_type_ == MeasuredTerminalType::shunt) {
+            return -1.0; // For shunt and load the direction in the math model is opposite to the direction in the
+                         // physical model
+        }
+        return 1.0;
+    }
+
+  private:
+    MeasuredTerminalType terminal_type_;
+    AngleMeasurementType angle_measurement_type_;
+
+    // virtual function getter for sym and asym param
+    // override them in real sensors function
+    virtual CurrentSensorCalcParam<symmetric_t> sym_calc_param() const = 0;
+    virtual CurrentSensorCalcParam<asymmetric_t> asym_calc_param() const = 0;
+
+    virtual CurrentSensorOutput<symmetric_t> get_sym_output(ComplexValue<symmetric_t> const& i) const = 0;
+    virtual CurrentSensorOutput<asymmetric_t> get_asym_output(ComplexValue<asymmetric_t> const& i) const = 0;
+};
+
+template <symmetry_tag current_sensor_symmetry_> class CurrentSensor : public GenericCurrentSensor {
+  public:
+    using current_sensor_symmetry = current_sensor_symmetry_;
+
+    static constexpr char const* name =
+        is_symmetric_v<current_sensor_symmetry> ? "sym_current_sensor" : "asym_current_sensor";
+    using InputType = CurrentSensorInput<current_sensor_symmetry>;
+    using UpdateType = CurrentSensorUpdate<current_sensor_symmetry>;
+    template <symmetry_tag sym_calc> using OutputType = CurrentSensorOutput<sym_calc>;
+
+    explicit CurrentSensor(CurrentSensorInput<current_sensor_symmetry> const& current_sensor_input, double u_rated)
+        : GenericCurrentSensor{current_sensor_input},
+          i_angle_measured_{current_sensor_input.i_angle_measured},
+          i_angle_sigma_{current_sensor_input.i_angle_sigma} {
+        set_current(current_sensor_input, u_rated);
+    };
+
+    UpdateChange update(CurrentSensorUpdate<current_sensor_symmetry> const& update_data, double const& u_rated) {
+        double const base_current = base_power_3p / u_rated / sqrt3;
+        double const scalar = convert_direction() / base_current;
+
+        if (!is_nan(update_data.i_sigma)) {
+            i_sigma_ = update_data.i_sigma / base_current;
+        }
+        if (!is_nan(update_data.i_angle_sigma)) {
+            i_angle_sigma_ = update_data.i_angle_sigma;
+        }
+        update_real_value<current_sensor_symmetry>(update_data.i_measured, i_measured_, scalar);
+        update_real_value<current_sensor_symmetry>(update_data.i_angle_measured, i_angle_measured_, 1.0);
+        return {false, false};
+    }
+
+    CurrentSensorUpdate<current_sensor_symmetry> inverse(CurrentSensorUpdate<current_sensor_symmetry> update_data,
+                                                         double const& u_rated) const {
+        assert(update_data.id == this->id() || is_nan(update_data.id));
+
+        double const base_current = base_power_3p / u_rated / sqrt3;
```

Comment:
Can be stored in base_i_ since it gets calculated multiple times

Reply:
Done